### PR TITLE
Removing Remote Entitlement

### PR DESCRIPTION
**Pull request by markusthoemmes**
This removes remote entitlement in its entirety and also removes the questionable "feature" of doing things cross-namespaces. Namespace switching is hereby explicit by setting the correct API key when authenticating with the OpenWhisk backend.

TBD:

- [ ] repurpose /namespaces to a "whoami" and tighten respective tests

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/controller/Namespaces.scala [hunk 1]
```diff
@@ -129,18 +113,8 @@ trait WhiskNamespacesApi
...
-                complete(OK, namespaces)
-            case Failure(r: RejectRequest) =>
-                logging.info(this, s"[GET] namespaces failed: ${r.message}")
-                terminate(r.code, r.message)
-            case Failure(t) =>
-                logging.error(this, s"[GET] namespaces failed: ${t.getMessage}")
-                terminate(InternalServerError)
-        }
+    private def getNamespaces(user: Identity)(implicit transid: TransactionId) = {
+        complete(OK, List(user.namespace))
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Namespaces.scala]: should this remain a list?

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 2]
```diff
@@ -84,12 +83,12 @@ abstract protected[controller] class RestAPIVersion(
 protected[controller] object RestAPIVersion_v1 {
     def requiredProperties =
         WhiskConfig.whiskVersion ++
-            WhiskServices.requiredProperties ++
+            WhiskAuthStore.requiredProperties ++
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: consolidating the required properties is ok - since the api unit tests use the common controller test trait which unifies the required properties also.

File: core/controller/src/main/scala/whisk/core/controller/Namespaces.scala (hunk 1 above)
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Namespaces.scala]: In the foreseeable future probably not. Didn't want too much of a semantic change in here though. That way, behaviour is at least theoretically preserved, as in: You can access every namespace in that list with that key.

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala (hunk 2 above)
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: I also thought it makes sense to have the properties required local to where the corresponding objects are created, which is in RestAPIs

File: tests/src/system/basic/WskBasicTests.scala [hunk 3]
```diff
@@ -688,9 +688,9 @@ class WskBasicTests
 
     behavior of "Wsk Namespace CLI"
 
-    it should "list namespaces" in {
+    it should "return a list of exactly one namespace" in {
         wsk.namespace.list().
-            stdout should include regex ("@|guest")
+            stdout.lines should have size 2 // headline + namespace
```
Comment (rabbah) [inline tests/src/system/basic/WskBasicTests.scala]: Can't we use wsk admin to retrieve what this value should be?

Comment (markusthoemmes) [inline tests/src/system/basic/WskBasicTests.scala]: No, `wskadmin` is not to be used in a system level test. Pressure on the `whoami` endpoint rises.

**Comment by rabbah** [comment]
Good catch.

**Comment by markusthoemmes** [comment]
PG3 311

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala [hunk 4]
```diff
@@ -221,23 +220,8 @@ protected[core] abstract class EntitlementProvider(config: WhiskConfig, loadBala
                 resource.collection.implicitRights(user, defaultNamespaces, right, resource) flatMap {
                     case true => Future.successful(true)
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: is this better as an if/else?

File: tests/src/system/basic/WskBasicTests.scala (hunk 3 above)
Comment (rabbah) [inline tests/src/system/basic/WskBasicTests.scala]: fair point.

File: tests/src/system/basic/WskBasicTests.scala [hunk 5]
```diff
@@ -728,7 +728,7 @@ class WskBasicTests
                     wsk.activation.get(activation.activationId, fieldFilter = Some("name")).stdout should include(s"""$successMsg name\n"$name"""")
                     wsk.activation.get(activation.activationId, fieldFilter = Some("version")).stdout should include(s"""$successMsg version\n"0.0.1"""")
                     wsk.activation.get(activation.activationId, fieldFilter = Some("publish")).stdout should include(s"""$successMsg publish\nfalse""")
-                    wsk.activation.get(activation.activationId, fieldFilter = Some("subject")).stdout should include regex (s"""(?i)$successMsg subject\n$ns_regex_list""")
```
Comment (rabbah) [inline tests/src/system/basic/WskBasicTests.scala]: why is this removed? `$ns_regex_lis`

Comment (markusthoemmes) [inline tests/src/system/basic/WskBasicTests.scala]: Because it's not true. The subject is not necessarily equal to the namespace used.

**Comment by junoyoon** [comment]
@markusthoemmes 

Is it planned to externalize the entitlement in the other way? 
We have been working on integrate openwhisk with [Keystone](https://docs.openstack.org/developer/keystone/)  using this feature.

**Comment by rabbah** [comment]
If we need to restore the entitlement host through whisk config to enable remote entitlement that's reasonable. Really what this cleanup was about was to remove some entanglement between the authorization and namespacing that belongs in a proper authorizarion layer and not the core. 

Once you have an integration with keystone, I suggest we look at what changes you have/are needed in the core and we can review that via a pr. Does that make sense?

**Comment by markusthoemmes** [comment]
It's also worth mentioning that our very own integration with Bluemix works completely asynchronous now. It only cares for the entries in the subjects database and is not entangled with the controller whatsoever.

Maybe you can achieve something similar with your Keystone integration?

**Comment by junoyoon** [comment]
@markusthoemmes @rabbah  
Thanks for explanation.  If I really need some changes on the core, I'll propose PR. :-)

**Comment by csantanapr** [comment]
Related to #1107
